devicetypes/smartthings/inovelli-dimmer.src/inovelli-dimmer.groovy

```diff
@@ -0,0 +1,528 @@
+		case "range":
+			return settings."$parameterKey"
+		default:
+			return Integer.parseInt(settings."$parameterKey")
+	}
+}
+
+private isNotNull(value) {
+	return value != null
+}
+
+private isPreferenceChanged(preference) {
+	if (isNotNull(settings."$preference.key")) {
+		if (preference.type == "boolRange") {
+			def boolName = preference.key + "Boolean"
+			if (state.currentPreferencesState."$preference.key".status == "disabled") {
+				return settings."$boolName"
+			} else {
+				return state.currentPreferencesState."$preference.key".value != settings."$preference.key" || !settings."$boolName"
+			}
+		} else {
+			return state.currentPreferencesState."$preference.key".value != settings."$preference.key"
+		}
+	} else {
+		return false
+	}
+}
+
+def parse(String description) {
+	def result = null
+	if (description != "updated") {
+		def cmd = zwave.parse(description)
+		if (cmd) {
+			result = zwaveEvent(cmd)
+			log.debug("'$description' parsed to $result")
+		} else {
+			log.debug("Couldn't zwave.parse '$description'")
+		}
+	}
+	log.debug "parsed '${description}' to ${result.inspect()}"
+	result
+}
+
+def handleLEDPreferenceEvent(cmd) {
+	def hueState = [name: "hue", value: "${Math.round(zwaveValueToHuePercent(cmd.scaledConfigurationValue))}"]
+	def childDni = "${device.deviceNetworkId}:1"
+	def childDevice = childDevices.find { it.deviceNetworkId == childDni }
+	childDevice?.sendEvent(hueState)
+	childDevice?.sendEvent(name: "saturation", value: "100")
+}
+
+def createChildDevice(childDthNamespace, childDthName, childDni, childComponentLabel, childComponentName) {
+	try {
+		log.debug "Creating a child device: ${childDthNamespace}, ${childDthName}, ${childDni}, ${childComponentLabel}, ${childComponentName}"
+		def childDevice = addChildDevice(childDthNamespace, childDthName, childDni, device.hub.id,
+			[
+				completedSetup: true,
+				label         : childComponentLabel,
+				isComponent   : true,
+				componentName : childComponentName,
+				componentLabel: childComponentLabel
+			])
+		childDevice
+	} catch (Exception e) {
+		log.debug "Exception: ${e}"
+	}
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.basicv1.BasicReport cmd) {
+	dimmerEvents(cmd)
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.switchmultilevelv3.SwitchMultilevelReport cmd) {
+	dimmerEvents(cmd)
+}
+
+def dimmerEvents(physicalgraph.zwave.Command cmd) {
+	log.debug "dimmer events ${cmd}"
+	def map = [name: "switch", value: cmd.value ? "on" : "off", descriptionText: "$device.displayName was turned $cmd.value"]
+	def switchEvent = createEvent(map)
+	def result = [switchEvent]
+	if (cmd.value) {
+		map = [name: "level", value: cmd.value == 99 ? 100 : cmd.value, unit: "%"]
+		result << createEvent(map)
+	}
+	if (switchEvent.isStateChange) {
+		result << response(["delay 3000", zwave.meterV3.meterGet(scale: 2).format()])
+	}
+	return result
+}
+
+def on() {
+	encapSequence([
+		zwave.basicV1.basicSet(value: 0xFF),
+		zwave.basicV1.basicGet()
+	], 5000)
```

Comment:
Is such a long delay really needed here (and in other places in this DTH) ?

Reply:
Done